server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedRareTerms.java

```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    protected void writeTermTypeInfoTo(StreamOutput out) throws IOException {
+        super.writeTermTypeInfoTo(out);
+        out.writeLong(maxDocCount);
+        bloom.writeTo(out);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        Map<Object, List<B>> buckets = new HashMap<>();
+        InternalTerms<A, B> referenceTerms = null;
+        BloomFilter bloomFilter = null;
+
+        for (InternalAggregation aggregation : aggregations) {
+
+            // Unmapped rare terms don't have a bloom filter so we'll skip all this work
+            // and save some type casting headaches later.
+            if (aggregation.isMapped() == false) {
+                continue;
+            }
+
+            @SuppressWarnings("unchecked")
+            InternalTerms<A, B> terms = (InternalTerms<A, B>) aggregation;
+            if (referenceTerms == null && aggregation.getClass().equals(UnmappedRareTerms.class) == false) {
+                referenceTerms = terms;
+            }
+            if (referenceTerms != null &&
+                referenceTerms.getClass().equals(terms.getClass()) == false &&
+                terms.getClass().equals(UnmappedRareTerms.class) == false) {
+                // control gets into this loop when the same field name against which the query is executed
+                // is of different types in different indices.
+                throw new AggregationExecutionException("Merging/Reducing the aggregations failed when computing the aggregation ["
+                    + referenceTerms.getName() + "] because the field you gave in the aggregation query existed as two different "
+                    + "types in two different indices");
+            }
+            for (B bucket : terms.getBuckets()) {
+                List<B> bucketList = buckets.computeIfAbsent(bucket.getKey(), k -> new ArrayList<>());
+                bucketList.add(bucket);
+            }
+
+            if (bloomFilter == null) {
+                bloomFilter = ((InternalMappedRareTerms)aggregation).bloom;
+            } else {
+                bloomFilter.merge(((InternalMappedRareTerms)aggregation).bloom);
+            }
+        }
+
+        // Always return all results, so just proactively size the array to num buckets
+        final int size = buckets.size();
+        final List<B> rare = new ArrayList<>(size);
+        for (List<B> sameTermBuckets : buckets.values()) {
+            final B b = sameTermBuckets.get(0).reduce(sameTermBuckets, reduceContext);
+            // Only prune if this is the final reduction, otherwise we may remove a term that shows
+            // up in a later incremental reduction and looks "rare" even though it isn't.
+            if (reduceContext.isFinalReduce() == false || (b.getDocCount() <= maxDocCount && containsTerm(bloom, b) == false)) {
+                rare.add(b);
+            }
```

Comment:
For the terms which merge and end up above the doc count limit do we not need to add them to the bloom filter? I think this is especially important for the case where there are multiple reduce phases (incremental reduce)

Reply:
Hmm, I think it's protected from that right now because we only "prune" the list if it's the final reduction, so we know there are no more terms and all same-term buckets are combined.

But that was added because there were issues with incremental reduction, and I think your proposal (add to bloom) would fix the issue without having to wait until the final reduction like I'm doing right now.

Will investigate, thanks :)